File: train.cpp
#include "include/tensor.h"
#include "include/ops.h"
#include "include/autograd.h"

int main() {
    // Initialize a simple dataset
    std::vector<float> input_data = {0.5, 1.5, -0.5, -1.5, 1.0, 2.0, -1.0, -2.0};
    std::vector<float> target_data = {1, 0, 1, 0};
    Tensor input({4, 2}, input_data, true);
    Tensor target({4}, target_data, true);

    // Initialize model and loss
    Linear linear1(2, 2);
    linear1.weights.data = {1, 1, 1, 1};
    linear1.bias.data = {0, 0};
    ReLU relu;
    Linear linear2(2, 2);
    linear2.weights.data = {1, 1, 1, 1};
    linear2.bias.data = {0, 0};
    CrossEntropy loss;

    // Create compute graph
    ComputeGraph graph;

    // Add operations to graph
    graph.add(&linear1);
    graph.add(&relu);
    graph.add(&linear2);
    graph.add(&loss);
    
    // Number of epochs
    int epochs = 10;

    // Training loop
    for (int epoch = 0; epoch < epochs; ++epoch) {
        // Forward pass
        Tensor output = linear1.forward(input);
        output = relu.forward(output);
        output = linear2.forward(output);
        Tensor loss_value = loss.forward(output, target);

        // Backward pass
        graph.backward();

        // Print loss
        std::cout << "Epoch " << epoch << ", Loss: " << loss_value.data[0] << std::endl;
        
        float learning_rate = 0.01;
        graph.sgd_step(learning_rate);
        // print weights and bias gradients
        std::cout << "Linear1 weights grad: " << linear1.weights.grad[0] << ", " << linear1.weights.grad[1] << ", " << linear1.weights.grad[2] << ", " << linear1.weights.grad[3] << std::endl;
        std::cout << "Linear1 bias grad: " << linear1.bias.grad[0] << ", " << linear1.bias.grad[1] << std::endl;
    }

    return 0;
}

File: train.py
import torch
import torch.nn as nn
import torch.optim as optim

# Initialize a simple dataset
input_data = torch.tensor([[0.5, 1.5], [-0.5, -1.5], [1.0, 2.0], [-1.0, -2.0]], dtype=torch.float32)
target_data = torch.tensor([1, 0, 1, 0], dtype=torch.long)

# Define the neural network model
class SimpleNN(nn.Module):
    def __init__(self):
        super(SimpleNN, self).__init__()
        self.linear1 = nn.Linear(2, 2)
        self.relu = nn.ReLU()
        self.linear2 = nn.Linear(2, 2)
        self._initialize_weights()

    def _initialize_weights(self):
        # Initialize weights to 1s and biases to 0s
        nn.init.constant_(self.linear1.weight, 1.0)
        nn.init.constant_(self.linear1.bias, 0.0)
        nn.init.constant_(self.linear2.weight, 1.0)
        nn.init.constant_(self.linear2.bias, 0.0)
    
    def forward(self, x):
        x = self.linear1(x)
        x = self.relu(x)
        x = self.linear2(x)
        return x

model = SimpleNN()

# Define the loss function and the optimizer
criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(model.parameters(), lr=0.01)

# Number of epochs
epochs = 10

# Training loop
for epoch in range(epochs):
    # Zero the parameter gradients
    optimizer.zero_grad()
    
    # Forward pass
    outputs = model(input_data)
    loss = criterion(outputs, target_data)
    
    # Backward pass
    loss.backward()
    
    # Optimize
    optimizer.step()
    
    # Print loss
    print(f"Epoch {epoch}, Loss: {loss.item()}")
    print(f"Gradients for linear1 weights: {model.linear1.weight.grad}")
    print(f"Gradients for linear1 bias: {model.linear1.bias.grad}")
